File: Leetcode/459.Repeated_Substring_Pattern/459.c
/* ************************************************************************
> File Name:     459.c
> Author:        Jack Yang
> Created Time:  2023年04月30日 (週日) 22時58分08秒
> Description:   Repeated Substring Pattern 
 ************************************************************************/
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<stdbool.h>
void next(char *s,int *nextTable){
    int len=strlen(s);
    int i;
    int j=0;
    nextTable[0]=j;
    for(i=1;i<len;i++){
        while(j>0 && s[i]!=s[j]){
            j=nextTable[j-1];
        }
        if(s[i]==s[j]){
            j++;
        }
        nextTable[i]=j;
    }
}
bool repeatedSubstringPattern(char * s){
    int len=strlen(s);
    int *nextTable=(int*)calloc(len,sizeof(int));
    int i;
    int j=0;
    next(s,nextTable);
    return nextTable[len-1]>0 && nextTable[len-1]%(len-nextTable[len-1])==0 ? true : false;
}

int main(int argc,char *argv[]){
    char s[]={'a','b','c','a','b','c','a','b','c','a','b','c','\0'};
    if (repeatedSubstringPattern(s))printf("true");
    else printf("false");
    return 0;
}

File: Leetcode/459. Repeated Substring Pattern/JavaScript/459.js
var repeatedSubstringPattern = function(s) {
  // 建立 next table
  function makeNextTable(s) {
    let j = 0;
    let next = [j];
    for (let i = 1 ; i < s.length ; i++) {
        while (j > 0 && s[j] != s[i]) {
            j = next[j - 1];
        }

        if (s[j] == s[i]) {
            j++;
        }

        next[i] = j;
    }

    return next;
  }  

  let next = makeNextTable(s);
  // next table index 的意義表示最長前後綴，
  // 當next table的最後一個值 > 0 表示最後一個值有最長前後綴，
  // 且當字串總長 - next table的最後一個值，表示可能重複的substring，
  // 因此當 next table的最後一個值 和 (字串總長 - next table的最後一個值) 的餘數 = 0，表示有連續重複substring
  return (next[s.length - 1] > 0 && (next[s.length - 1] % (s.length - next[s.length - 1]) == 0)) ? 1 : 0;
};

let s = "abcabcabcabc";
console.log(repeatedSubstringPattern(s));